File: LeetCode/Combinations.java
public class Solution {
    public int numOnes(int n) {
        int ans = 0;
        for(int bit = 1; bit != bit << 1; bit <<= 1)
            if((n & bit) != 0)
                ans++;
        return ans;
    }
    
    public ArrayList<ArrayList<Integer>> combine(int n, int k) {
        ArrayList<ArrayList<Integer>> ans = new ArrayList<ArrayList<Integer>>();
        for(int i = 0; i < (1<<n); i++) {
            ArrayList<Integer> a = new ArrayList<Integer>();
            if(numOnes(i) == k)
                for(int j = 0; j < n; j++)
                    if(((i >> j) & 1) != 0)
                        a.add(i + 1);
            ans.add(a);
        }
        return ans;
    }
}

File: LeetCode/Combinations.cc
typedef vector<int> vi;
typedef vector<vi> vii;

class Solution {
public:
    map<pair<int,int>,vii> m;
    vector<vector<int> > combine(int n, int k) {
        return c(n, k);
    }
    const vii &c(int n, int k) {
        if(m.count(make_pair(n, k))) return m[make_pair(n, k)];
        if(n < k) return m[make_pair(n, k)] = vii();
        if(k == 0) return m[make_pair(n, k)] = vii(1, vi());
        vii ans(c(n-1, k));
        const vii &v = c(n-1, k-1);
        for(vi vv : v) {
            vv.push_back(n);
            ans.push_back(vv);
        }
        return m[make_pair(n, k)] = ans;
    }
};